src/box/raft.c

```diff
@@ -462,27 +465,54 @@ box_raft_try_promote_f(struct trigger *trig, void *event)
 	struct raft *raft = event;
 	assert(raft == box_raft());
 	struct box_raft_watch_ctx *ctx = trig->data;
+
+	if (raft->state == RAFT_STATE_LEADER)
+		goto done;
+
 	/*
 	 * Term ended with nothing, probably split vote which led to a next
 	 * term.
 	 */
-	if (raft->volatile_term > ctx->term)
+	if (raft->volatile_term > ctx->term) {
+		diag_set(ClientError, ER_OLD_TERM,
+			 (unsigned long long)ctx->term,
+			 (unsigned long long)raft->volatile_term);
 		goto done;
+	}
+
+	int healthy_count = replicaset.healthy_count;
+	if (healthy_count < replicaset_healthy_quorum()) {
+		diag_set(ClientError, ER_NO_ELECTION_QUORUM,
+			 healthy_count, replicaset_healthy_quorum());
+		goto done;
+	}
+
 	/*
 	 * Instance does not participate in terms as a candidate anymore. Can
 	 * happen not only if the node was a temporary candidate and the term
 	 * was bumped, but also if it was reconfigured during the waiting or it
 	 * lost the connection quorum.
 	 */
-	if (!raft->is_candidate)
+	if (!raft->is_candidate) {
+		if (box_election_mode == ELECTION_MODE_MANUAL) {
+			diag_set(TimedOut);
+		} else {
+			assert(box_election_mode != ELECTION_MODE_CANDIDATE);
+			diag_set(ClientError, ER_ELECTION_DISABLED);
+		}
 		goto done;
-	/* The term ended with a leader being found. */
-	if (raft->leader != REPLICA_ID_NIL)
+	}
+
+	if (raft->leader != 0) {
```

Comment:
Nit: it was `if (raft->leader != REPLICA_ID_NIL)`, no need to change, if it's not really needed